### PR TITLE
Complete Netherite Implementation

### PR DESCRIPTION
**Pull request by LoboMetalurgico**
### Before we can merge this pull request, we need to fulfill these items:

**GET ID:**

- [x] Netherite Block - 525

- [x] Netherite Scrap - 752
- [x] Netherite Ingot - 742

- [x] Netherite Sword - 743
- [x] Netherite Pickaxe - 745
- [x] Netherite Shovel - 744
- [x] Netherite Axe - 746
- [x] Netherite Hoe - 747

- [x] Netherite Helmet - 748
- [x] Netherite Chestplate - 749
- [x] Netherite Leggings - 750
- [x] Netherite Boots - 751

Thanks @good777LUCKY for give me the necessary id's :)

**GET COLOR:**

- [x] Netherite Block Color

**SOUND:**

- [x] Netherite Sound Armor Equip (ID)

- [x] Fix Hoe Sound

- [x] Fix Shovel Sound

**ENCHANT ABILITY:**

- [x] Netherite Armors - 15

- [x] Netherite Tools - 15

**VERIFY:**

- [x] Netherite Armor Toughness - 3

- [x] Max Block Id

**MECHANICS:**

- [x] Implement Fast Leave Break using Hoe

- [x] Implement Fast Soul Sand Break using Hoe - Not On Bedrock :/

### CODE REVIEW

**Comment by LoboMetalurgico** [comment]
### Before we can merge this pull request, we need to fulfill these items:

**GET ID:**

- [ ] Netherite Block

- [ ] Netherite Scrap
- [ ] Netherite Ingot

- [ ] Netherite Sword
- [ ] Netherite Pickaxe
- [ ] Netherite Shovel
- [ ] Netherite Axe
- [ ] Netherite Hoe

- [ ] Netherite Helmet
- [ ] Netherite Chestplate
- [ ] Netherite Leggings
- [ ] Netherite Boots

**GET COLOR:**

- [ ] Netherite Block Color

**SOUND:**

- [ ] Netherite Sound Armor Equip (ID)

**ENCHANT ABILITY:**

- [ ] Netherite Armors

- [ ] Netherite Tools

**Comment by good777LUCKY** [comment]
**IDs:**
- Netherite Block: 525
- Netherite Scrap: 752
- Netherite Ingot: 742
- Netherite Sword: 743
- Netherite Pickaxe: 745
- Netherite Shovel: 744
- Netherite Axe: 746
- Netherite Hoe: 747
- Netherite Helmet: 748
- Netherite Chestplate: 749
- Netherite Leggings: 750
- Netherite Boots: 751

**Comment by LoboMetalurgico** [comment]
Thank You! :)

**Comment by good777LUCKY** [comment]
Does Max Block ID means max block id of whole blocks in Bedrock Edition?

**Comment by joserobjr** [comment]
@good777LUCKY no, it determine the size of the internal arrays. PowerNukkit's storage supports up to 65535 block ids

**Comment by joserobjr** [comment]
It also used to support block data value up to 63 (6 bits), but this was increased to in this PR to 4,294,967,295 (32 bits) when the block needs to use the new hyper-id system

**Comment by good777LUCKY** [comment]
There was no specific sound for Netherite Armor Equip in Bedrock Edition Vanilla assets. Only netherite armor sound is missing...

**Comment by good777LUCKY** [comment]
I found reason why there was no netherite armor equip sound. It was not implemented until 1.16.20.50 beta.
Screenshot
(1.16.20.50 beta changelog)

**Comment by good777LUCKY** [comment]
1.16.20.50 beta's sound ID was `item.armor.equip_netherite`.

**Comment by LoboMetalurgico** [comment]
Thank You Again!

**Comment by zombihunterbonn** [comment]
How long you net for finish it? @joserobjr

**Comment by Jl1mbo** [comment]
You can find out where you get **getHandler()**, **getResistance()** and **getColor()** ?

**Comment by joserobjr** [comment]
@Jl1mbo they usually match with the Java Edition implementation, so we can see that information using FabricMC

**Comment by Jl1mbo** [comment]
> @Jl1mbo they usually match with the Java Edition implementation, so we can see that information using FabricMC

Please link :)

**Comment by joserobjr** [comment]
@Jl1mbo https://fabricmc.net/ an example project: https://github.com/FabricMC/fabric-example-mod (it's usually enough to see what's inside the latest Minecraft Java Edition using the IDE)

**Comment by Superzock3003** [comment]
When I place the Netherite Block, it becomes this Block (screenshot) and that is in the console "[ERROR] Found an unknown BlockId: Meta combination: 525: 0, replacing with an" UPDATE! "Block."
Minecraft 02 07 2020 10_18_47

**Comment by Jl1mbo** [comment]
> When I place the Netherite Block, it becomes this Block (screenshot) and that is in the console "[ERROR] Found an unknown BlockId: Meta combination: 525: 0, replacing with an" UPDATE! "Block."
> Minecraft 02 07 2020 10_18_47
> 

+

File: src/main/java/cn/nukkit/block/BlockNetherite.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        return 6000;
+    }
+
+    @Override
+    public Item[] getDrops(Item item) {
+        if (item.isPickaxe() && item.getTier() >= ItemTool.TIER_DIAMOND) {
+            return new Item[]{
+                    toItem()
+            };
+        } else {
```
Comment (HeyZeer0) [inline src/main/java/cn/nukkit/block/BlockNetherite.java:42]: there's no need for placing an else here

File: src/main/java/cn/nukkit/block/BlockGrassPath.java [hunk 2]
```diff
@@ -57,6 +59,9 @@ public int onUpdate(int type) {
     @Override
     public boolean onActivate(Item item, Player player) {
         if (item.isHoe()) {
+            if(player != null){
+                player.getLevel().addSound(player, Sound.USE_GRASS);
```
Comment (HeyZeer0) [inline src/main/java/cn/nukkit/block/BlockGrassPath.java:63]: Probably better to play the sound AFTER the action is executed for consistent responses

File: src/main/java/cn/nukkit/block/BlockGrass.java [hunk 3]
```diff
@@ -61,10 +62,16 @@ public boolean onActivate(Item item, Player player) {
...
         } else if (item.isHoe()) {
+            if(player != null){
+                player.getLevel().addSound(player, Sound.USE_GRASS);
+            }
             item.useOn(this);
             this.getLevel().setBlock(this, Block.get(BlockID.FARMLAND));
             return true;
         } else if (item.isShovel()) {
+            if(player != null){
+                player.getLevel().addSound(player, Sound.USE_GRASS);
```
Comment (HeyZeer0) [inline src/main/java/cn/nukkit/block/BlockGrass.java:73]: Probably better to play the sound AFTER the action is executed for consistent responses

File: src/main/java/cn/nukkit/block/BlockGrass.java [hunk 4]
```diff
@@ -61,10 +62,16 @@ public boolean onActivate(Item item, Player player) {
             ObjectTallGrass.growGrass(this.getLevel(), this, new NukkitRandom());
             return true;
         } else if (item.isHoe()) {
+            if(player != null){
+                player.getLevel().addSound(player, Sound.USE_GRASS);
```
Comment (HeyZeer0) [inline src/main/java/cn/nukkit/block/BlockGrass.java:66]: Probably better to play the sound AFTER the action is executed for consistent responses

File: src/main/java/cn/nukkit/block/BlockDirt.java [hunk 5]
```diff
@@ -54,6 +56,9 @@ public String getName() {
     @Override
     public boolean onActivate(Item item, Player player) {
         if (item.isHoe()) {
+            if(player != null){
+                player.getLevel().addSound(player, Sound.USE_GRASS);
```
Comment (HeyZeer0) [inline src/main/java/cn/nukkit/block/BlockDirt.java:60]: Probably better to play the sound AFTER the action is executed for consistent responses

File: src/main/java/cn/nukkit/block/Block.java [hunk 6]
```diff
@@ -872,7 +883,11 @@ private static double speedRateByHasteLore0(int hasteLoreLevel) {
         return 1.0 + (0.2 * hasteLoreLevel);
     }
 
-    private static int toolType0(Item item) {
+    @PowerNukkitDifference
+    private static int toolType0(Item item, int blockId, int blockToolType) {
+        if((blockId == LEAVES && item.isHoe()) ||
+                (blockId == LEAVES2 && item.isHoe()))
+            return 5;
```
Comment (HeyZeer0) [inline src/main/java/cn/nukkit/block/Block.java:890]: Probably better to follow the format and use the ItemTool class for easier readability

Comment (LoboMetalurgico) [reply src/main/java/cn/nukkit/block/Block.java:890]: The ItemTool class doesn't have a simple way to make this work. After several tests, I saw that this was the best way to resolve this issue, considering that this is where the tool verification is done, and not in the ItemTool class.

However, I will do the proper formatting.

File: src/main/java/cn/nukkit/block/BlockNetherite.java (hunk 1 above)
Comment (LoboMetalurgico) [reply src/main/java/cn/nukkit/block/BlockNetherite.java:42]: The else was used on this occasion to maintain the pattern followed by the other blocks.

**Comment by Jl1mbo** [comment]
Why are these blocks still not loaded in PowerNukkit?

**Comment by joserobjr** [comment]
> Why are these blocks still not loaded in PowerNukkit?

Depending on #341

File: src/main/java/cn/nukkit/item/ItemArmor.java [hunk 7]
```diff
@@ -17,7 +18,8 @@
     public static final int TIER_CHAIN = 3;
     public static final int TIER_GOLD = 4;
     public static final int TIER_DIAMOND = 5;
-    public static final int TIER_OTHER = 6;
+    public static final int TIER_NETHERITE = 6;
+    public static final int TIER_OTHER = 7;
```
Comment (joserobjr) [inline src/main/java/cn/nukkit/item/ItemArmor.java:22]: Unnecessary breaking change (changing the constant value of `TIER_OTHER `)

**Pull request merged**